Add Koharu

KenjieDec opened the pull request:
Closes #3951

Checklist:
- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

KenjieDec committed:
Add Koharu.to

KenjieDec committed:
Remove some useless things

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,279 @@
+    override val supportsLatest = true
+
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder()
+        .rateLimit(3, 1, TimeUnit.SECONDS)
```
```suggestion
        .rateLimit(1)
```
harsher rate limit because they are known to be problematic

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,279 @@
+        .rateLimit(3, 1, TimeUnit.SECONDS)
+        .build()
+
+    private val json = Json { ignoreUnknownKeys = true }
```
```suggestion
    private val json: Json by injectLazy()

```

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,279 @@
+        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
+    }
+
+    private var quality = preferences.getString(PREF_IMAGERES, "0")!!
```
```suggestion
    private fun quality() = preferences.getString(PREF_IMAGERES, "0")!!
```

also set a resampled quality as default, which they use on the site

KenjieDec replied:
Hello, I can't seem to find any resampled resolution in their API json. Should I just default the value to just 1280?

AwkwardPeak7 replied:
whichever quality they use by default on the site, i know they dont use original

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,279 @@
+                9 -> females.add(tag.name + " ♀")
+                10 -> mixed.add(tag.name + " ◊")
+                12 -> other.add(tag.name + " ◊")
+                else -> tags.add(tag.name + " ◊")
```
what is this symbol? imo no need to add?

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,279 @@
+    private fun readInstanceProperty(instance: Data, propertyName: String): Any? {
+        val property = instance::class.members.first { it.name == propertyName } as KProperty1<Any, *>
+        return property.get(instance)
+    }
```
![thisisfine](https://cdn.discordapp.com/emojis/927962828294471681.webp?size=48&quality=lossless)

~you can use java reflection which avoids adding another library?~ check discord #dev channel

KenjieDec committed:
Fixed Filters, and More

- Fixed Filters/Search not working when query is empty
- Changed image resolution default value to match website's
- Use `when` instead of kotlin's reflect
- Added search by "id", and url intent filter?

vetleledaal commented on src/en/koharu/AndroidManifest.xml:
```diff
@@ -0,0 +1,23 @@
+            </intent-filter>
+        </activity>
+    </application>
+</manifest>
```
```suggestion
</manifest>

```

vetleledaal commented on src/en/koharu/build.gradle:
```diff
@@ -0,0 +1,8 @@
+    isNsfw = true
+}
+
+apply from: "$rootDir/common.gradle"
```
```suggestion
apply from: "$rootDir/common.gradle"

```

vetleledaal commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,303 @@
+                    else -> {}
+                }
+            }
+            addQueryParameter("s", "title:\"$query \" " + terms.joinToString(" "))
```
If terms is empty it will have a trailing space. Might be used for fingerprinting.

KenjieDec committed:
Apply Suggestions

- Apply vetleledaal's suggestions
- Fix UrlActivity?

KenjieDec committed:
Apply suggestions

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,304 @@
+
+    companion object {
+        const val PREFIX_ID_KEY_SEARCH = "id:"
+        private const val PREF_IMAGERES = "1280"
```
```suggestion
        private const val PREF_IMAGERES = "pref_image_quality"

```

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/Koharu.kt:
```diff
@@ -0,0 +1,304 @@
+        val data = entry.data
+        val dataKey = when (quality()) {
+            "1600" -> data.`1600`
+            "1280" -> data.`980`
```
980?

AwkwardPeak7 commented on src/en/koharu/src/eu/kanade/tachiyomi/extension/en/koharu/KoharuDto.kt:
```diff
@@ -0,0 +1,75 @@
+    val `780`: DataKey = `0`,
+    val `980`: DataKey = `780`,
+    val `1280`: DataKey = `980`,
+    val `1600`: DataKey = `1280`,
```
why assign one previous value as default?

KenjieDec replied:
Their API sometimes doesn't have a specific resolution, like 1600x, so I set the previous value as the default to be safe. 
Should I handle it in a different way?

AwkwardPeak7 replied:
should be fine but reverse the order, so if user chooses 780 and it is not available use 980 instead of original, and so on

KenjieDec replied:
Hello, sorry for late commit. Can you please check the new code and see if it matches what you asked for?

AwkwardPeak7 replied:
I meant just changing the default assignments order, but I suppose this works too

KenjieDec committed:
Apply Suggestions

KenjieDec committed:
Add files via upload
